File: src/services/scheduler.py
"""Scheduler service for automated weekly roulette execution."""

import asyncio
import threading
from datetime import datetime, time
from zoneinfo import ZoneInfo
from typing import Dict, List

import schedule

from database.connection import get_database
from database.models import ChannelConfig
from .roulette import RouletteService


class SchedulerService:
    """Service for scheduling and running automated roulettes."""

    def __init__(self, roulette_service: RouletteService):
        """Initialize scheduler service."""
        self.roulette_service = roulette_service
        self.db = get_database()
        self._running = False
        self._thread: threading.Thread = None
        self._scheduled_jobs: Dict[str, object] = {}
        self.poland_tz = ZoneInfo("Europe/Warsaw")
        self.utc_tz = ZoneInfo("UTC")

    def _convert_polish_time_to_utc(self, time_str: str) -> str:
        """Convert Polish time to UTC for scheduling.

        Args:
            time_str: Time in HH:MM format (Polish time)

        Returns:
            Time in HH:MM format (UTC)
        """
        try:
            hour, minute = map(int, time_str.split(':'))

            today = datetime.now(self.poland_tz).date()
            polish_time = datetime.combine(today, time(hour, minute), tzinfo=self.poland_tz)

            utc_time = polish_time.astimezone(self.utc_tz)

            return utc_time.strftime("%H:%M")

        except Exception as e:
            print(f"⚠️ Error converting time {time_str} to UTC: {e}")
            return time_str

    def _schedule_channel_roulette(self, config: ChannelConfig) -> None:
        """Schedule roulette for a specific channel configuration."""
        if not config.enabled:
            return

        def run_channel_roulette():
            """Run roulette for this specific channel."""
            print(f"🎯 Running scheduled roulette for channel {config.channel_id}")
            try:
                loop = asyncio.new_event_loop()
                asyncio.set_event_loop(loop)
                result = loop.run_until_complete(
                    self.roulette_service.send_roulette_message(
                        config.channel_id, test_mode=False
                    )
                )
                loop.close()

                if result["success"]:
                    print(
                        f"✅ Roulette completed for {config.channel_id}: {result['selected_member']['name']}"
                    )
                else:
                    print(
                        f"❌ Roulette failed for {config.channel_id}: {result['error']}"
                    )
            except Exception as e:
                print(f"💥 Error running roulette for {config.channel_id}: {e}")

        job_key = f"{config.channel_id}_{config.day}_{config.time}"

        if job_key in self._scheduled_jobs:
            schedule.cancel_job(self._scheduled_jobs[job_key])

        utc_time = self._convert_polish_time_to_utc(config.time)

        job = (
            getattr(schedule.every(), config.day)
            .at(utc_time)
            .do(run_channel_roulette)
        )
        self._scheduled_jobs[job_key] = job

        print(
            f"📅 Scheduled roulette for channel {config.channel_id}: {config.day} at {config.time} PL ({utc_time} UTC)"
        )

    def update_schedules(self) -> None:
        """Update all schedules based on current database configuration."""
        print("🔄 Updating roulette schedules...")

        schedule.clear()
        self._scheduled_jobs.clear()

        configs = self.db.get_all_enabled_configs()

        if not configs:
            print("ℹ️ No enabled roulette configurations found")
            return

        for config in configs:
            self._schedule_channel_roulette(config)

        print(f"✅ Updated schedules for {len(configs)} channels")

    def start(self) -> None:
        """Start the scheduler in a background thread."""
        if self._running:
            print("⚠️ Scheduler is already running")
            return

        self._running = True
        self.update_schedules()

        def run_scheduler():
            """Run the scheduler loop."""
            print("🚀 Starting scheduler thread...")
            while self._running:
                schedule.run_pending()
                threading.Event().wait(60)
            print("🛑 Scheduler thread stopped")

        self._thread = threading.Thread(target=run_scheduler, daemon=True)
        self._thread.start()
        print("✅ Scheduler started successfully")

    def stop(self) -> None:
        """Stop the scheduler."""
        if not self._running:
            return

        print("🛑 Stopping scheduler...")
        self._running = False
        schedule.clear()
        self._scheduled_jobs.clear()

        if self._thread and self._thread.is_alive():
            self._thread.join(timeout=5)

        print("✅ Scheduler stopped")

    def get_schedule_info(self) -> List[dict]:
        """Get information about all scheduled jobs."""
        jobs_info = []

        for job_key, job in self._scheduled_jobs.items():
            parts = job_key.split("_")
            if len(parts) >= 3:
                channel_id = "_".join(parts[:-2])
                day = parts[-2]
                time_str = parts[-1]

                jobs_info.append(
                    {
                        "channel_id": channel_id,
                        "day": day,
                        "time": time_str,
                        "next_run": job.next_run.isoformat() if job.next_run else None,
                        "job_key": job_key,
                    }
                )

        return jobs_info

    def force_update_schedules(self) -> None:
        """Force update schedules (useful when configuration changes)."""
        if self._running:
            self.update_schedules()
        else:
            print("⚠️ Scheduler is not running, schedules will be updated when started")

File: internal/services/scheduler.go
package services

import (
	"context"
	"fmt"
	"log/slog"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/go-co-op/gocron"
	"github.com/typeWolffo/weeklyroulette-bot/internal/database"
)

type SchedulerService struct {
	scheduler *gocron.Scheduler
	roulette  *RouletteService
	db        *database.Repository
	mu        sync.RWMutex
	jobs      map[string]*gocron.Job
	polandTZ  *time.Location
	running   bool
}

func NewSchedulerService(roulette *RouletteService, db *database.Repository) (*SchedulerService, error) {
	polandTZ, err := time.LoadLocation("Europe/Warsaw")
	if err != nil {
		return nil, fmt.Errorf("loading timezone: %w", err)
	}

	s := gocron.NewScheduler(time.UTC)

	return &SchedulerService{
		scheduler: s,
		roulette:  roulette,
		db:        db,
		jobs:      make(map[string]*gocron.Job),
		polandTZ:  polandTZ,
	}, nil
}

func (s *SchedulerService) convertPolishTimeToUTC(timeStr string) (string, error) {
	parts := strings.Split(timeStr, ":")
	if len(parts) != 2 {
		return timeStr, fmt.Errorf("invalid time format: %s", timeStr)
	}

	hour, err := strconv.Atoi(parts[0])
	if err != nil {
		return timeStr, fmt.Errorf("invalid hour: %s", parts[0])
	}

	minute, err := strconv.Atoi(parts[1])
	if err != nil {
		return timeStr, fmt.Errorf("invalid minute: %s", parts[1])
	}

	now := time.Now().In(s.polandTZ)
	polishTime := time.Date(now.Year(), now.Month(), now.Day(), hour, minute, 0, 0, s.polandTZ)
	utcTime := polishTime.UTC()

	return fmt.Sprintf("%02d:%02d", utcTime.Hour(), utcTime.Minute()), nil
}

func dayToWeekday(day string) time.Weekday {
	switch strings.ToLower(day) {
	case "monday":
		return time.Monday
	case "tuesday":
		return time.Tuesday
	case "wednesday":
		return time.Wednesday
	case "thursday":
		return time.Thursday
	case "friday":
		return time.Friday
	case "saturday":
		return time.Saturday
	case "sunday":
		return time.Sunday
	default:
		return time.Monday
	}
}

func (s *SchedulerService) scheduleChannelRoulette(config *database.ChannelConfig) error {
	if !config.Enabled {
		return nil
	}

	utcTime, err := s.convertPolishTimeToUTC(config.Time)
	if err != nil {
		slog.Warn("failed to convert time to UTC", "time", config.Time, "error", err)
		utcTime = config.Time
	}

	jobKey := fmt.Sprintf("%s_%s_%s", config.ChannelID, config.Day, config.Time)

	channelID := config.ChannelID
	job, err := s.scheduler.Every(1).Week().Weekday(dayToWeekday(config.Day)).At(utcTime).Do(func() {
		slog.Info("running scheduled roulette", "channel", channelID)
		ctx := context.Background()

		result, err := s.roulette.SendRouletteMessage(ctx, channelID, false)
		if err != nil {
			slog.Error("scheduled roulette failed", "channel", channelID, "error", err)
			return
		}

		if result.Success {
			slog.Info("roulette completed", "channel", channelID, "selected", result.SelectedMember.Name)
		} else {
			slog.Error("roulette failed", "channel", channelID, "error", result.Error)
		}
	})

	if err != nil {
		return fmt.Errorf("scheduling roulette: %w", err)
	}

	s.mu.Lock()
	s.jobs[jobKey] = job
	s.mu.Unlock()

	slog.Info("scheduled roulette",
		"channel", config.ChannelID,
		"day", config.Day,
		"polish_time", config.Time,
		"utc_time", utcTime,
	)

	return nil
}

func (s *SchedulerService) UpdateSchedules() {
	slog.Info("updating roulette schedules")

	s.mu.Lock()
	s.scheduler.Clear()
	s.jobs = make(map[string]*gocron.Job)
	s.mu.Unlock()

	ctx := context.Background()
	configs, err := s.db.GetAllEnabledConfigs(ctx)
	if err != nil {
		slog.Error("failed to get enabled configs", "error", err)
		return
	}

	if len(configs) == 0 {
		slog.Info("no enabled roulette configurations found")
		return
	}

	for _, config := range configs {
		if err := s.scheduleChannelRoulette(config); err != nil {
			slog.Error("failed to schedule channel", "channel", config.ChannelID, "error", err)
		}
	}

	slog.Info("updated schedules", "count", len(configs))
}

func (s *SchedulerService) Start() {
	s.mu.Lock()
	if s.running {
		s.mu.Unlock()
		slog.Warn("scheduler is already running")
		return
	}
	s.running = true
	s.mu.Unlock()

	s.UpdateSchedules()
	s.scheduler.StartAsync()
	slog.Info("scheduler started")
}

func (s *SchedulerService) Stop() {
	s.mu.Lock()
	defer s.mu.Unlock()

	if !s.running {
		return
	}

	slog.Info("stopping scheduler")
	s.scheduler.Stop()
	s.running = false
	s.jobs = make(map[string]*gocron.Job)
	slog.Info("scheduler stopped")
}

func (s *SchedulerService) ForceUpdateSchedules() {
	s.mu.RLock()
	running := s.running
	s.mu.RUnlock()

	if running {
		s.UpdateSchedules()
	} else {
		slog.Warn("scheduler is not running, schedules will be updated when started")
	}
}

func (s *SchedulerService) GetScheduleInfo() []map[string]interface{} {
	s.mu.RLock()
	defer s.mu.RUnlock()

	var jobsInfo []map[string]interface{}

	for jobKey, job := range s.jobs {
		parts := strings.Split(jobKey, "_")
		if len(parts) >= 3 {
			channelID := strings.Join(parts[:len(parts)-2], "_")
			day := parts[len(parts)-2]
			timeStr := parts[len(parts)-1]

			var nextRun string
			if nextRunTime := job.NextRun(); !nextRunTime.IsZero() {
				nextRun = nextRunTime.Format(time.RFC3339)
			}

			jobsInfo = append(jobsInfo, map[string]interface{}{
				"channel_id": channelID,
				"day":        day,
				"time":       timeStr,
				"next_run":   nextRun,
				"job_key":    jobKey,
			})
		}
	}

	return jobsInfo
}
